base/task.jl

```diff
@@ -250,7 +250,9 @@ true
 """
 istaskfailed(t::Task) = (load_state_acquire(t) === task_state_failed)
 
-Threads.threadid(t::Task) = Int(ccall(:jl_get_task_tid, Int16, (Any,), t)+1)
+Threads.threadid(t::Task) = Int(ccall(:jl_get_task_relative_tid, Int16, (Any,), t)+1)
+Threads.rawthreadid(t::Task) = Int(ccall(:jl_get_task_tid, Int16, (Any,), t)+1)
```

Comment:
I think `nthreads()` should be the total number of worker threads in all thread pools and `threadid() in 1:nthreads()`. That's what you care for writing correct and efficient low-level concurrent algorithms that work for on an arbitrary thread pool. For example, it's required for implementing multi-queue in the user space and guarantee lock-free enqueue/dequeue on an arbitrary worker thread.

I don't *think* this approach introduces any bugs to correctly written code using `@spawn`. It might waste more memory but they just have to fix it by not using `threadid` which we recommend these days anyway. I guess the only "bugs" that can be introduced by this is when people do `@threads :static for i in 1:nthreads()` and rely on the `:static` scheduling for the forward progress (e.g., spin inside the loop body). But I'd consider it less bad than introducing subtle races. It's also kind of a terrible approach to parallelism anyway and we get the chance to say "please don't do this."

Reply:
> when people do `@threads :static for i in 1:nthreads()`

Moving forward, I wonder if we can just expose a simple API `Threads.foreachworker(f, pool_name)` where `f` gets the relative id and called on a sticky task on each worker thread. That's what people actually need when they use `@threads` this way.

Reply:
This would almost certainly break Dagger.jl, for example, which relies on every thread in `nthreads()` to be safely scheduled onto: https://github.com/JuliaParallel/Dagger.jl/blob/0e0379ba885e31afb9683d037af1027c79ff651a/src/Dagger.jl#L89 and https://github.com/JuliaParallel/Dagger.jl/blob/0e0379ba885e31afb9683d037af1027c79ff651a/src/processor.jl#L167 would cause unintended usage of threadpools by Dagger. While one can definitely argue that this code is bad because it's calling `jl_set_task_tid` manually, I do expect that this pattern is used elsewhere.

Basically, my hope is that once placed in a threadpool, code should be entirely unaware of other threadpools, and local initialization of per-thread buffers should be done based on the locally-available threads. Buffer initialization at the global level (at `__init__` time) will have to be adjusted to account for total threads in the session, and instead use `Threads.rawthreadid()` to access such buffers, which I think is fair given that they're trying to be non-local and need to have explicit knowledge of the size of all threadpools.

Reply:
> While one can definitely argue that this code is bad because it's calling `jl_set_task_tid` manually,

As I said, the PR as-is breaks other classes of code. Furthermore, the code that will be broken by this PR can be implemented without touching any internals. I believe it's clear that we should prefer to avoid breaking code using public APIs and not code `ccall`ing the runtime internal `jl_set_task_tid`, if they are the only choices.

>  I do expect that this pattern is used elsewhere.

I do wish it's not used *everywhere* and people still have some hope in the vision of composable parallel libraries.